pallets/corporate-actions/src/distribution.rs

```diff
@@ -0,0 +1,555 @@
+
+        /// Has an asset holder been paid yet?
+        ///
+        /// (CAId, DID) -> Was DID paid in the CAId?
+        HolderPaid get(fn holder_paid):
+            double_map hasher(blake2_128_concat) CAId, hasher(blake2_128_concat) IdentityId => bool;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Start and attach a capital distribution, to the CA identified by `ca_id`,
+        /// with `amount` funds in `currency` withdrawn from `portfolio` belonging to `origin`'s DID.
+        ///
+        /// The distribution will commence at `payment_at` and expire at `expires_at`,
+        /// if provided, or if `None`, then there's no expiry.
+        ///
+        /// ## Arguments
+        /// - `origin` which must be a signer for the CAA of `ca_id`.
+        /// - `ca_id` identifies the CA to start a capital distribution for.
+        /// - `portfolio` specifies the portfolio number of the CAA to distribute `amount` from.
+        /// - `currency` to withdraw and distribute from the `portfolio`.
+        /// - `amount` of `currency` to withdraw and distribute.
+        /// - `payment_at` specifies when shares may first be pushed or claimed.
+        /// - `expires_at` speciies, if provided, when remaining shares are forfeit
+        ///    and may be reclaimed by `origin`.
+        ///
+        /// # Errors
+        /// - `UnauthorizedAsAgent` if `origin` is not `ticker`'s sole CAA (owner is not necessarily the CAA).
+        /// - `DistributingAsset` if `ca_id.ticker == currency`.
+        /// - `ExpiryBeforePayment` if `expires_at.unwrap() <= payment_at`.
+        /// - `NowAfterPayment` if `payment_at < now`.
+        /// - `NoSuchCA` if `ca_id` does not identify an existing CA.
+        /// - `NoRecordDate` if CA has no record date.
+        /// - `RecordDateAfterStart` if CA's record date > payment_at.
+        /// - `InsufficientPortfolioBalance` if `portfolio` has less than `amount` of `currency`.
+        /// - `InsufficientBalance` if the protocol fee couldn't be charged.
+        #[weight = 2_000_000_000]
+        pub fn distribute(
+            origin,
+            ca_id: CAId,
+            portfolio: PortfolioNumber,
+            currency: Ticker,
+            amount: T::Balance,
+            payment_at: Moment,
+            expires_at: Option<Moment>,
+        ) {
+            // Ensure CA's asset is distinct from the distributed currency.
+            ensure!(ca_id.ticker != currency, Error::<T>::DistributingAsset);
+
+            // Ensure that any expiry date doesn't come before the payment date.
+            ensure!(!expired(expires_at, payment_at), Error::<T>::ExpiryBeforePayment);
+
+            // Ensure `now <= payment_at`.
+            ensure!(<Checkpoint<T>>::now_unix() <= payment_at, Error::<T>::NowAfterPayment);
```

Comment:
I believe `now <= start` is generally consistent with what we use elsewhere, e.g., `record_date <= start`.
Moreover, there's no harm with being inclusive re. current time rather than requiring a strict relation.

Reply:
I think then you have to make the same change for the `remove_distribution()` pallet. Where we are not allowing to remove at an inclusive time. I think as per the definition of `payment_at` user should able to withdraw its funds when `now == payment_at`.

Reply:
I don't think that follows. However, there's actually a bug (fixing) in `ensure_active_distribution`; should have been `now >= dist.payment_at`, not `now > dist.payment_at`, which is (part of) the definition of an active distribution. You should then only be able to remove a distribution if it isn't active yet, i.e. the complement of `now >= dist.payment_at`, i.e. `!(now >= dist.payment_at)` => `now < dist.payment_at`.